spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,279 @@
+	public void setTimeout(Duration timeout) {
+		this.timeout = timeout;
+	}
+
+	/**
+	 * A {@code boolean} indicating if the {@link BigQueryFileMessageHandler} should synchronously
+	 * wait for each file to be successfully loaded to BigQuery.
+	 * @param sync whether {@link BigQueryFileMessageHandler} should wait synchronously for jobs to
+	 * 		complete
+	 */
+	public void setSync(boolean sync) {
+		this.sync = sync;
+	}
+
+	@Override
+	protected ListenableFuture handleRequestMessage(Message<?> message) {
+		String datasetName =
+				message.getHeaders().containsKey(BigQuerySpringMessageHeaders.DATASET_NAME)
+						? message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class)
+						: this.datasetNameExpr.getValue(this.evaluationContext, message, String.class);
+
+		String tableName =
+				message.getHeaders().containsKey(BigQuerySpringMessageHeaders.TABLE_NAME)
+						? message.getHeaders().get(BigQuerySpringMessageHeaders.TABLE_NAME, String.class)
+						: this.tableNameExpr.getValue(this.evaluationContext, message, String.class);
+
+		FormatOptions formatOptions =
+				message.getHeaders().containsKey(BigQuerySpringMessageHeaders.FORMAT_OPTIONS)
+						? message.getHeaders().get(
+								BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class)
+						: this.formatOptions;
+
+		TableId tableId = TableId.of(datasetName, tableName);
+
+		WriteChannelConfiguration writeChannelConfiguration = WriteChannelConfiguration
+				.newBuilder(tableId)
+				.setFormatOptions(formatOptions)
+				.setAutodetect(this.autoDetectSchema)
+				.setWriteDisposition(this.writeDisposition)
+				.build();
+
+		TableDataWriteChannel writer = bigQuery.writer(writeChannelConfiguration);
+
+		try (
+				InputStream inputStream = convertToInputStream(message.getPayload());
+				OutputStream sink = Channels.newOutputStream(writer)) {
+			// Write data from data input file to BigQuery
+			StreamUtils.copy(inputStream, sink);
+		}
+		catch (FileNotFoundException e) {
+			throw new MessageHandlingException(
+					message, "Failed to find file to write to BigQuery in message handler: " + this, e);
+		}
+		catch (IOException e) {
+			throw new MessageHandlingException(
+					message, "Failed to write data to BigQuery tables in message handler: " + this, e);
+		}
+
+		if (writer.getJob() == null) {
+			throw new MessageHandlingException(
+					message, "Failed to initialize the BigQuery write job in message handler: " + this);
+		}
+
+		// Prepare the polling task for the ListenableFuture result returned to end-user
+		SettableListenableFuture result = new SettableListenableFuture();
+		ScheduledFuture<?> scheduledFuture = taskScheduler.scheduleAtFixedRate(() -> {
+			Job job = writer.getJob().reload();
+			if (State.DONE.equals(job.getStatus().getState())) {
+				if (job.getStatus().getError() != null) {
+					result.setException(
+							new RuntimeException(job.getStatus().getError().getMessage()));
+				}
+				else {
+					result.set(job);
```

Comment:
What is the business purpose of this `Job`? Why do we really need it in comparison of sent file?

Reply:
So Job is useful because the user can get statistics and metadata from the load operation out of it. For example # of successfully inserted rows, # of unsuccessfully loaded rows, time taken, etc.

https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/bigquery/Job.html